### PR TITLE
Release/version11 7 4

### PR DESCRIPTION
**Pull request by otanum**
1. MediaData is by design InvariantCulture in Optimizely. Moved to separate index. Searches with FileSearchProvider returns hits for all editor languages.
2. Added support for neutral and specific cultures. Note that indexname and language use new split char. Indices must be recreated. Delete old.
3. Adjusted primary key for tracking table to support multiple languages. Combined primary key for indexname, language and search word.
4. Fixed some best bets issues. Light adjustments for admin GUI.
5. Added ExcludeField

### CODE REVIEW

File: src/Epinova.ElasticSearch.Core.EPiServer/Controllers/Abstractions/ElasticSearchControllerBase.cs [hunk 1]
```diff
@@ -71,33 +77,36 @@ protected override void OnResultExecuting(ResultExecutingContext filterContext)
...
-                return new
                 {
-                    Index = nameWithoutLanguage,
-                    i.DisplayName
-                };
-            })
-            .Distinct()
-            .ToDictionary(x => x.Index, x => x.DisplayName);
+                    string lang = _settings.GetLanguageFromIndexName(i.Index);
+                    //var nameWithoutLanguage = _settings.GetIndexNameWithoutLanguage(i.Index);
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/Controllers/Abstractions/ElasticSearchControllerBase.cs:96]: Kan den fjernes?

File: src/Epinova.ElasticSearch.Core.EPiServer/Controllers/Abstractions/ElasticSearchControllerBase.cs [hunk 2]
```diff
@@ -111,7 +120,7 @@ protected override void Initialize(RequestContext requestContext)
 
         private List<IndexInformation> GetIndices()
         {
-            var indexHelper = new Admin.Index(_serverInfoService, _settings, _httpClientHelper, "GetIndices-NA");
+            var indexHelper = new Admin.Index(_serverInfoService, _settings, _httpClientHelper, null);
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/Controllers/Abstractions/ElasticSearchControllerBase.cs:123]: Når man sitter og leser kode som her i nettleser, er det til stor nytte å prefikse med navn på argument for null, tall, bool osv: Altså fortell meg hva som er navnet på argumentet som får verdien null.

File: src/Epinova.ElasticSearch.Core.EPiServer/EpiserverSearchLanguage.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                if(!_enabledLanguages.Any())
+                    return CultureInfo.InvariantCulture;
+
+                return GetEnabledCulture(CultureInfo.CurrentCulture) ?? GetEnabledCulture(CultureInfo.CurrentCulture.Parent) ?? GetEnabledCulture(CultureInfo.CurrentCulture.Parent.Parent) ?? CultureInfo.InvariantCulture;
+            }
+        }
+        
+        private CultureInfo GetEnabledCulture(CultureInfo cultureInfo)
+        {
+            return _enabledLanguages.FirstOrDefault(l => l.Equals(cultureInfo));
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/EpiserverSearchLanguage.cs:36]: Is SingleOrDefault more precise?

File: src/Epinova.ElasticSearch.Core.EPiServer/Extensions/CustomResultExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+                _trackingRepository.AddSearch(service, noHits: results.TotalHits == 0);
+
+            return results;
+        }
+
+        public static CustomSearchResult<T> GetCustomResults<T>(this IElasticSearchService<T> service)
+        {
+            CustomSearchResult<T> results = service.GetResultsCustom();
+            
+            if(service.TrackSearch && !string.IsNullOrWhiteSpace(service.IndexName))
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/Extensions/CustomResultExtensions.cs:26]: Consolidate the essence with duplicated code in async version

File: src/Epinova.ElasticSearch.Core.EPiServer/Initialization/TrackingInitializer.cs [hunk 5]
```diff
@@ -26,9 +29,26 @@ public void Initialize(InitializationEngine context)
...
+            {
+                DbHelper.AdjustColumnNullable(connectionString, Constants.Tracking.TableName, Constants.TrackingFieldIndex, "nvarchar(200)", isNullable: false);
+            }
 
-                DbHelper.ExecuteCommand(connectionString, $"ALTER TABLE {Constants.Tracking.TableName} ADD {Constants.TrackingFieldIndex} nvarchar(200)");
-                DbHelper.ExecuteCommand(connectionString, $"IF (OBJECT_ID('PK_ElasticTracking', 'U') IS NOT NULL) BEGIN ALTER TABLE {Constants.Tracking.TableName}  DROP CONSTRAINT [PK_ElasticTracking] END");
+            FixPrimaryKeys(connectionString);
+        }
+
+        private void FixPrimaryKeys(string connectionString)
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/Initialization/TrackingInitializer.cs:42]: 🙈

File: src/Epinova.ElasticSearch.Core.EPiServer/Models/ViewModels/ConsoleViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+
+namespace Epinova.ElasticSearch.Core.EPiServer.Models.ViewModels
+{
+    public class ConsoleViewModel
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/Models/ViewModels/ConsoleViewModel.cs:5]: Could this be simplified to record?

File: src/Epinova.ElasticSearch.Core.EPiServer/Models/ViewModels/SettingsViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections.Generic;
+
+namespace Epinova.ElasticSearch.Core.EPiServer.Models.ViewModels
+{
+    public class SettingsViewModel
```
Comment (tholee) [inline src/Epinova.ElasticSearch.Core.EPiServer/Models/ViewModels/SettingsViewModel.cs:5]: Did you consider record?

Comment (otanum) [reply src/Epinova.ElasticSearch.Core.EPiServer/Models/ViewModels/SettingsViewModel.cs:5]: Not this time

**Pull request merged**